### PR TITLE
[24.1] Fix empty usernames in database + bug in username generation

### PR DESCRIPTION
**Pull request by jdavcs**
This is a follow-up to #18364 (ref https://github.com/galaxyproject/galaxy/pull/18364#issuecomment-2159846131).
I've testing this locally (the CI will run the migration too)

This also fixes a newly-discovered bug introduced in #16724 during migrating db code to SA2.0 (see commit description).
I'll backport the bug fix to 23.2. I'm not targeting this at 23.2 because this has tests that rely on data access testing infrastructure introduced in 24.1.

Question: do we want to apply the same type of fix to the tool_shed database schema and data? If so, I'll add a commit with a TS migration.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by jdavcs** [comment]
The packages error is an import error (`managers` modules are not available for `data`). I'll fix this.

**Comment by jdavcs** [comment]
I have dropped the unit test for now ([it passed](https://github.com/galaxyproject/galaxy/actions/runs/9474061191/job/26102916703)) - I'll add it in a separate PR targeting dev. For this to be testable from within the data package, the test should not import from managers. In dev, this is easily solvable by moving db-specific functions into the new `model/db` location.

File: lib/galaxy/model/migrations/alembic/versions_gxy/c63848676caf_update_username_column_schema_and_data.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+"""
+
+from alembic import op
+from sqlalchemy import (
+    or_,
+    select,
+    update,
+)
+
+from galaxy.managers.users import username_from_email_with_connection
```
Comment (mvdbeek) [inline lib/galaxy/model/migrations/alembic/versions_gxy/c63848676caf_update_username_column_schema_and_data.py]: Can't import from managers here, that breaks the package structure. Generally speaking it's probably wise to have a fixed copy in the migration so we don't alter migration results accidentally.

**Review by mvdbeek — APPROVED** [review]
Thank you!

**Pull request merged**